### PR TITLE
Implementing Gradient Accumulation Optimizer

### PR DESCRIPTION
**Pull request by stefan-falk**
# Description

This PR is a fork/alternative of https://github.com/tensorflow/addons/pull/2525 and seeks to implement gradient accumulation.

## Type of change

- [ ] Bug fix
- [ ] New Tutorial
- [ ] Updated or additional documentation
- [ ] Additional Testing
- [ ] New Activation and the changes conform to the [activation contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/activations/README.md#contribution-guidelines)
- [ ] New Callback and the changes conform to the [callback contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/callbacks/README.md#contribution-guidelines)
- [ ] New Image addition and the changes conform to the [image op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/image/README.md#contribution-guidelines)
- [ ] New Layer and the changes conform to the [layer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/layers/README.md#contribution-guidelines)
- [ ] New Loss and the changes conform to the [loss contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/losses/README.md#contribution-guidelines)
- [ ] New Metric and the changes conform to the [metric contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/metrics/README.md#contribution-guidelines)
- [x] New Optimizer and the changes conform to the [optimizer contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/README.md#contribution-guidelines)
- [ ] New RNN Cell and the changes conform to the [rnn contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/rnn/README.md#contribution-guidelines)
- [ ] New Seq2seq addition and the changes conform to the [seq2seq contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/seq2seq/README.md#contribution-guidelines)
- [ ] New Text addition and the changes conform to the [text op contribution guidelines](https://github.com/tensorflow/addons/blob/master/tensorflow_addons/text/README.md#contribution-guidelines)

# Checklist:

- [ ] I've properly [formatted my code according to the guidelines](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md#coding-style)
    - [ ] By running Black + Flake8
    - [ ] By running pre-commit hooks
- [ ] This PR addresses an already submitted issue for TensorFlow Addons
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] This PR contains modifications to C++ custom-ops

### CODE REVIEW

File: tensorflow_addons/optimizers/tests/gradient_accumulator_test.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+#     config = opt.get_config()
+#
+#     assert config["accum_steps"] == accum_steps
+#
+#     new_opt = GradientAccumulator.from_config(config)
+#     old_sgd_config = opt._optimizer.get_config()
+#     new_sgd_config = new_opt._optimizer.get_config()
+#
+#     for k1, k2 in zip(old_sgd_config, new_sgd_config):
+#         assert old_sgd_config[k1] == new_sgd_config[k2]
```
Comment (stefan-falk) [inline tensorflow_addons/optimizers/tests/gradient_accumulator_test.py:138]: See https://github.com/tensorflow/addons/pull/2527/files#r672857887

**Comment by stefan-falk** [comment]
@fsx950223 With this implementation I do not get any errors on my side. The (ASR) model I seek to train appears to converge.

**Comment by fsx950223** [comment]
> @fsx950223 With this implementation I do not get any errors on my side. The (ASR) model I seek to train appears to converge.

Do you know the reason?

**Comment by stefan-falk** [comment]
> Do you know the reason?

The reason why it works? No idea 😆

**Comment by stefan-falk** [comment]
I can only assume that it's because internally I never have any branching or something like that 🤷‍♂️ 

I'll admit that I am not really an expert for the internal workings of Tensorflow (yet).

But if you want I can make this PR on your forked repository on https://github.com/fsx950223/addons/tree/ga ?

**Comment by stefan-falk** [comment]
@googlebot I consent.

File: tensorflow_addons/optimizers/tests/gradient_accumulator_test.py [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+    accu_steps = 2
+    opt = GradientAccumulator(
+        tf.keras.optimizers.SGD(lr=1.0),
+        accu_steps=accu_steps,
+        trainable_variables=variables,
+    )
+    for _ in range(accu_steps * 4):
+        opt.apply_gradients(grads_and_vars)
+    np.testing.assert_allclose(
+        var0.read_value(), [[1.0, 2.0, 0.0], [0.2, 1.2, 0.0]], rtol=1e-5
```
Comment (fsx950223) [inline tensorflow_addons/optimizers/tests/gradient_accumulator_test.py:78]: remove rtol=1e-5

File: tensorflow_addons/optimizers/tests/gradient_accumulator_test.py [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+
+    grads_and_vars = list(zip([grads0, grads1], [var0, var1]))
+
+    variables = [var for _, var in grads_and_vars]
+
+    opt = GradientAccumulator(
+        tf.keras.optimizers.SGD(lr=1.0), accum_steps, trainable_variables=variables
+    )
+
+    for _ in range(accum_steps + 1):
```
Comment (fsx950223) [inline tensorflow_addons/optimizers/tests/gradient_accumulator_test.py:44]: accum_steps*2+1

File: tensorflow_addons/optimizers/tests/gradient_accumulator_test.py (hunk 2 above)
Comment (stefan-falk) [reply tensorflow_addons/optimizers/tests/gradient_accumulator_test.py:78]: I'm not sure why but the result has some rounding errors. I need to lower the tolerance here for the test to succeed. The gradients are computed correctly but the variables are updated with rounding errors.

File: tensorflow_addons/optimizers/tests/gradient_accumulator_test.py (hunk 3 above)
Comment (stefan-falk) [reply tensorflow_addons/optimizers/tests/gradient_accumulator_test.py:44]: You suggest to change this to:

```suggestion
    for _ in range(accum_steps * 2 + 1):
```

?

**Comment by fsx950223** [comment]
Close temporarily for deduplicate notes.
